core/src/main/java/io/grpc/internal/MessageSender.java

```diff
@@ -0,0 +1,293 @@
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.Compressor;
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+import java.io.InputStream;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * A class that manages the outbound state on behalf of a stream. Manages the sending of headers,
+ * messages, and trailers, as well as outbound flow control.
+ */
+final class MessageSender {
+  /**
+   * The default number of queued bytes for a given stream, below which {@link
+   * StreamListener#onReady()} will be called.
+   */
+  private static final int ONREADY_THRESHOLD = 32 * 1024;
+
+  private enum Phase {
+    HEADERS, MESSAGE, STATUS
+  }
+
+  /**
+   * The handler for this state.
+   */
+  interface Handler {
+    /**
+     * Sends the given headers to the remote endpoint.
+     */
+    void sendHeaders(Metadata headers);
+
+    /**
+     * Sends the given message to the remote endpoint.
+     *
+     * @param frame the message to be sent.
+     * @param endOfStream if {@code true}, indicates that the frame should be marked as the end of stream.
+     * @param flush if {@code true}, indicates that any buffered IO should be flushed to the remote endpoint.
+     */
+    void sendMessage(@Nullable WritableBuffer frame, boolean endOfStream, boolean flush);
+
+    /**
+     * Sends the specified trailers to the remote endpoint. This is only called for server-side
+     * handlers.
+     *
+     * @param trailers the trailers to be sent
+     * @param headersSent if {@code true}, indicates that headers were previously sent.
+     */
+    void sendTrailers(Metadata trailers, boolean headersSent);
```

Comment:
If this maintains the outbound state, shouldn't it know if the headers have been sent already?